bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/SerialIoThread.java

```diff
@@ -0,0 +1,284 @@
+        } catch (final InterruptedException e) {
+            // ignore
+        }
+    }
+
+    @Override
+    public void run() {
+        // RXTX serial port library causes high CPU load
+        // Start event listener, which will just sleep and slow down event loop
+        try {
+            serialPort.addEventListener(this);
+        } catch (final TooManyListenersException e) {
+            logger.warn("serial port setup failed", e);
+            return;
+        }
+        serialPort.notifyOnDataAvailable(true);
+
+        final byte[] buffer = new byte[256];
+        try (final InputStream in = serialPort.getInputStream()) {
+            enterMessageMode();
+            int len;
+            while (!done && (len = in.read(buffer)) >= 0) {
+                addData(buffer, len);
+            }
+        } catch (final Exception e) {
+            logger.warn("Exception in UPB read thread", e);
+        } finally {
+            logger.debug("shutting down receive thread");
+            shutdownAndAwaitTermination(writeExecutor);
+            serialPort.removeEventListener();
+            try {
+                serialPort.close();
+            } catch (final Exception e) {
+                // ignore
+            }
+        }
+        logger.debug("UPB read thread stopped");
+    }
+
+    private void addData(final byte[] data, final int length) {
+        if (bufferLength + length > buffer.length) {
+            // buffer overflow, discard entire buffer
+            bufferLength = 0;
+        }
+        System.arraycopy(data, 0, buffer, bufferLength, length);
+        bufferLength += length;
+        interpretBuffer();
+    }
+
+    private int findMessageLength(final byte[] buffer, final int bufferLength) {
+        for (int i = 0; i < bufferLength; i++) {
+            if (buffer[i] == 13) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Attempts to interpret any messages that may be contained in the buffer.
+     */
+    private void interpretBuffer() {
+        int messageLength = findMessageLength(buffer, bufferLength);
+
+        while (messageLength != -1) {
+            final String message = new String(Arrays.copyOfRange(buffer, 0, messageLength), US_ASCII);
+            logger.debug("UPB Message: {}", message);
+
+            final int remainingBuffer = bufferLength - messageLength - 1;
+            if (remainingBuffer > 0) {
+                System.arraycopy(buffer, messageLength + 1, buffer, 0, remainingBuffer);
+            }
+            bufferLength = remainingBuffer;
+            handleMessage(UPBMessage.fromString(message));
+            messageLength = findMessageLength(buffer, bufferLength);
+        }
+    }
+
+    private void handleMessage(final UPBMessage msg) {
+        switch (msg.getType()) {
+            case ACK:
+                if (currentWrite != null) {
+                    currentWrite.ackReceived(true);
+                }
```

Comment:
You should create a local variable for `currentWrite` to prevent threading issues and to satisfy the compile time null checker.